backends/dpdk/dpdkArch.cpp

```diff
@@ -2448,4 +2446,86 @@ const IR::Node* ElimHeaderCopy::postorder(IR::Member* m) {
     return m;
 }
 
+const IR::Node* DoFlattenHeaderUnion::postorder(IR::Type_Struct* s) {
+    IR::IndexedVector<IR::StructField> fields;
+    for (auto sf : s->fields) {
+        auto ftype = typeMap->getType(sf, true);
+        if (ftype->is<IR::Type_HeaderUnion>()) {
+            for (auto sfu : ftype->to<IR::Type_HeaderUnion>()->fields) {
+                cstring uName = sf->name.name + "_" + sfu->name.name;
+                auto uType = sfu->type->getP4Type();
+                fields.push_back(new IR::StructField(IR::ID(uName), uType));
+            }
+        } else {
+            fields.push_back(sf);
+        }
+    }
+    return new IR::Type_Struct(s->name, s->annotations, fields);
+}
+
+const IR::Node* DoFlattenHeaderUnion::postorder(IR::Declaration_Variable *dv) {
+    auto ftype = typeMap->getTypeType(dv->type, true);
+    if (ftype->is<IR::Type_HeaderUnion>()) {
+        for (auto sfu : ftype->to<IR::Type_HeaderUnion>()->fields) {
+             cstring uName = dv->name.name + "_" + sfu->name.name;
+             auto uType =  sfu->type->getP4Type();
+             toInsert.push_back(new IR::Declaration_Variable(IR::ID(uName), uType));
+        }
+    }
+    return dv;
+}
+
+const IR::Node* DoFlattenHeaderUnion::postorder(IR::Member* m) {
+    if (m->expr->type->to<IR::Type_HeaderUnion>()) {
+        if (auto huf = m->expr->to<IR::Member>()) {
+            return new IR::Member(huf->expr,IR::ID(huf->member.name+"_"+m->member.name));
```

Comment:
this is not guaranteed to generate a field name that does not exist.
the safer way would be to use the NameGenerator.
Then you will have to carry a map of the renamed fields around.

Reply:
Ok. I will update the code to use name generator/refmap.